tensorflow-ops/src/TensorFlow/Gradient.hs

```diff
@@ -710,6 +710,27 @@ opGrad "Pad" _ [toT -> x, toT -> padPattern] [dz] =
     gradientSliceBegin = CoreOps.reshape padPatternSliced rankx
     gradientSliceSize = shape (x :: Tensor Build Float)
 
+-- Gradient for Slice
+-- Create an Nx2 padding where N ist the rank of (grad of) Slice and the first
+-- column represents how many zeros are to be prepended for each dimension, and the second
+-- column indicates how many zeros are appended.
+-- The number of zeros to prepend is the shape of the beginvec.
+-- The number of zeros to append is the shape of the inputvec
+-- elementwise-subtracted by both the beginvec and sizevec.
+-- Some more reshaping is needed to assemble this tensor with the
+-- right dimensions.
+opGrad "Slice" _ [toT -> inputvec, toT -> beginvec, _] [dz] =
+   [Just $ CoreOps.pad dz paddings, Nothing, Nothing]
+  where
+    v1 = vector [1 :: Int32]
+    input_rank' = CoreOps.rank (inputvec :: Tensor Build Float)
+    -- For some reason input_rank' has an empty shape
```

Comment:
The rank is a scalar. Scalars are distinct from vectors of size one, so instead of shape=[1] and rank=1, they have shape=[] and rank=0.

I think you want

```haskell
input_rank = CoreOps.rank (inputvec :: Tensor Build Float)
pad_shape = CoreOps.concat 0 [input_rank, scalar 1]
```

Reply:
Yes, I want something like this, but you can't concatenate something with an empty shape. I tried to use the code above, but it gives the following error message:

_ERROR: TensorFlowException TF_INVALID_ARGUMENT "ConcatOp : Expected concatenating dimensions in the range [0, 0), but got 0\n\t [[Node: Concat_14 = Concat[N=2, T=DT_INT32, _device=\"/job:localhost/replica:0/task:0/device:CPU:0\"](Const_12, Rank_13, Const_11)]]"_

Reply:
Ah, that is surprising. The python API must be doing something to compensate for this.